regen/lock_definitions.pl

```diff
@@ -0,0 +1,2401 @@
+#       other such functions, or the functions in k) that have the same tag.
+#       Continuing the previous example, the function putenv() has
+#       "const:env".  This means it needs an exclusive lock on the mutex
+#       associated with "env", and all functions that contain just "env" for
+#       their data need read-locks on that mutex.
+#
+#   m)  The string "race".  This means that these each of these uses has
+#       a potential race with something running in another thread.  If "race"
+#       appears alone, what the other thing(s) that can interfere with it are
+#       unspecified, but the generated header takes it as meaning the use
+#       only has a race with another instance of it.  This could be because of
+#       a buffer shared between threads, or simply that it returns a pointer
+#       to internal global static storage, which must be used while still
+#       locked, or copied to a per-thread safe place for later use.
+#
+#       "race" may be followed by a colon and a tag word, like "race:tmpbuf".
+#       The potential race is with any other uses that also specify a
+#       race with the same tag word.
+#
+#       The implication is that each such use must be protected with an
+#       exclusive mutex associated with that tag, so none can run
+#       concurrently.
+#
+#       A condition may be attached to a race, as in "race:mbrlen/!ps".  The
+#       condition is introduced by a single '/'.  This means that the race
+#       doesn't happen unless the condition is met.  If you look at the
+#       mbrlen() man page, you will find that it takes an argument named "ps".
+#       What the condition tells us is that any call to mbrlen with "!ps",
+#       (hence ps is NULL) is thread-unsafe.  You can easily write your code
+#       so that "ps" is non-NULL, and remove this cause of unsafety.  The
+#       generated macros assume that you do so.
+#
+#   n)  A string giving a locale category, like "LC_TIME".  This indicates
+#       what locale category affects the execution of this use.  Multiple
+#       ones may be specified.  These are for future use. XXX
+#
+#   o)  The string "//".  This must be the final field on the line, and
+#       the rest of the line becomes a comment string that is to be output
+#       just above the generated macros for each affected use.  Any
+#       continuation lines continue this comment.
+#
+# There is another type of continuation line.  If the last non-blank character
+# in the uses column is a comma, there are more uses to come.  These
+# come on the lines immediately following any data column continuation lines.
+# These lines are simply more comma-separated use names.  The final line
+# doesn't end in a comma.
+#
+# The #endif preprocessor line marks the end of the uses affected by
+# previous preprocessor lines.  If there was no "#else" line, macros expanding
+# to no-ops are automatically generated for the #else case.
+
+__DATA__
```

Comment:
Two issues, in general:

1) This defines a huge number of names most of which will never be used, and have a potential to conflict with names from other libraries.

I think this should add a `PERL_` prefix to each lock name.

2) This defines separate symbols for each function in a closely related group of functions, for example `dbm_*` and any read locale function.

The separate macros imply the need to do each of those locks, even though in some cases it doesn't make sense, eg. if I do a dbm_delete(), I really don't want to release the lock until I do `dbm_error()`.

I think macros should be defined for groups of such functions where it makes sense, eg. a single macro for the `dbm_*` functions, and just document the need for `LC_*_LOCK` for locale reading functions etc.

Reply:
I share @tonycoz's skepticism as to whether the addition of this header file is necessary or desirable.

Reply:
@khwilliamson can you comment on the issues raised by @tonycoz in this ticket back in June?

(My sense is that there's not enough enthusiasm about the patch to warrant keeping this ticket open.)

Reply:
I think it could be useful, but it needs some work.

Reply:
You can close it, but I will open it again when I finally get around to fixing it up

Reply:
Sounds like a plan.  Closing ticket for the time being.